### ISSUE TITLE
Military Base Housing Price Prediction

### ISSUE THREAD

**Issue by ciaome2** [ciaome2/jaeseong-lee-project#1]
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

def main():
    # 1. Load the data
    train = pd.read_csv('./data/train.csv')
    test = pd.read_csv('./data/test.csv')
    
    print("Train data shape:", train.shape)
    print("Test data shape:", test.shape)
    print("First few rows of training data:\n", train.head())
    
    # 2. Handle missing values in 'total_bedrooms' by replacing them with the median value
    median_bedrooms = train['total_bedrooms'].median()
    train['total_bedrooms'].fillna(median_bedrooms, inplace=True)
    test['total_bedrooms'].fillna(median_bedrooms, inplace=True)
    
    # 3. One-Hot Encode the categorical variable 'ocean_proximity'
    print("Unique values in ocean_proximity:", train['ocean_proximity'].unique())
    
    # Use OneHotEncoder with drop='first' to avoid multicollinearity
    encoder = OneHotEncoder(sparse=False, drop='first')
    ocean_proximity_train = encoder.fit_transform(train[['ocean_proximity']])
    ocean_proximity_test = encoder.transform(test[['ocean_proximity']])
    
    # Get encoded column names
    encoded_cols = encoder.get_feature_names_out(['ocean_proximity'])
    
    # Convert the encoded arrays into DataFrames
    ocean_proximity_train_df = pd.DataFrame(ocean_proximity_train, columns=encoded_cols, index=train.index)
    ocean_proximity_test_df = pd.DataFrame(ocean_proximity_test, columns=encoded_cols, index=test.index)
    
    # Drop the original 'ocean_proximity' column and concatenate the new one-hot encoded columns
    train = pd.concat([train.drop('ocean_proximity', axis=1), ocean_proximity_train_df], axis=1)
    test = pd.concat([test.drop('ocean_proximity', axis=1), ocean_proximity_test_df], axis=1)
    
    print("First few rows after encoding:\n", train.head())
    
    # 4. Separate features and target variable from the training data
    X = train.drop('median_house_value', axis=1)
    y = train['median_house_value']
    
    # Remove the 'id' column from test data
    X_test = test.drop('id', axis=1)
    
    # 5. Scale numerical features using StandardScaler
    numerical_features = ['housing_median_age', 'total_rooms', 'total_bedrooms',
                          'population', 'households', 'median_income']
    
    scaler = StandardScaler()
    scaler.fit(X[numerical_features])
    
    X[numerical_features] = scaler.transform(X[numerical_features])
    X_test[numerical_features] = scaler.transform(X_test[numerical_features])
    
    # 6. Split the training data into training and validation sets
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # 7. Hyperparameter tuning using GridSearchCV on RandomForestRegressor
    rf = RandomForestRegressor(random_state=42)
    
    param_grid = {
        'n_estimators': [100, 200, 300],
        'max_depth': [None, 10, 20, 30],
        'min_samples_split': [2, 5, 10],
        'min_samples_leaf': [1, 2, 4],
        'bootstrap': [True, False]
    }
    
    grid_search = GridSearchCV(
        estimator=rf,
        param_grid=param_grid,
        cv=3,
        n_jobs=-1,
        verbose=2,
        scoring='neg_root_mean_squared_error'
    )
    
    grid_search.fit(X_train, y_train)
    
    print("Best hyperparameters:", grid_search.best_params_)
    
    best_rf = grid_search.best_estimator_
    
    # 8. Evaluate the best model on the validation set
    y_pred = best_rf.predict(X_val)
    rmse = np.sqrt(mean_squared_error(y_val, y_pred))
    print("Validation RMSE:", rmse)
    
    # 9. Predict on the test set and create the submission file
    test_pred = best_rf.predict(X_test)
    
    submission = pd.DataFrame({
        'id': test['id'],
        'median_house_value': test_pred
    })
    
    submission.to_csv('submission.csv', index=False)
    print("submission.csv file has been saved.")

if __name__ == '__main__':
    main()
